<!doctype html>
<html lang="en">
 <head>
  <meta charset="UTF-8">
  <title>Moving from one catch to another try catch block</title>
  <link rel="stylesheet" type="text/css" href="/css/base.css">
  <script type="text/x-mathjax-config">
            MathJax.Hub.Config({"HTML-CSS": { preferredFont: "TeX", availableFonts: ["STIX","TeX"], linebreaks: { automatic:true }, EqnChunk: (MathJax.Hub.Browser.isMobile ? 10 : 50) },
            tex2jax: { inlineMath: [ ["$", "$"], ["\\\\(","\\\\)"] ], displayMath: [ ["$$","$$"], ["\\[", "\\]"] ], processEscapes: true, ignoreClass: "tex2jax_ignore|dno" },
            TeX: {
            extensions: ["begingroup.js"],
            noUndefined: { attributes: { mathcolor: "red", mathbackground: "#FFEEEE", mathsize: "90%" } },
            Macros: { href: "{}" }
            },
            messageStyle: "none",
            styles: { ".MathJax_Display, .MathJax_Preview, .MathJax_Preview > *": { "background": "inherit" } },
            SEEditor: "mathjaxEditing"
            });
        </script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS_HTML-full"></script>
 </head>
 <body>
  <div class="question">
   <a href="/questions">All Questions</a>
   <h2>Moving from one catch to another try catch block</h2>
   <div class="container-metadata">
    <div>
     <span>Score: </span> <span>3</span>
    </div>
    <div>
     <span>Asker: </span> <span>Kaneki</span>
    </div>
    <div>
     <span>Asked: </span> <span>26 Jun 2020 at 04:16</span>
    </div>
    <div>
     <a href="https://stackoverflow.com/questions/62587952/moving-from-one-catch-to-another-try-catch-block">source</a>
    </div>
   </div>
   <div>
    <p>I have multiple try catch blocks. I want to make my program so that if something isn't found in the first block, the catch of that block will continue it to the other set of try catch. I did some research but wasn't able to come up with a solution.</p>
    <pre><code>try{
    Something
} catch (Exception e) {
    Move to next set of try catch
    }
</code></pre>
    <p>This is what I want to achieve.</p>
   </div>
   <div class="tags">
    <span class="tag">java</span>
   </div>
   <hr>
   <div class="comment">
    <table>
     <tbody>
      <tr>
       <td></td>
       <td><span>Why do you think that you need to nest the try-catch blocks like this?</span> <span> - </span> <span class="display-name">Tim Biegeleisen</span> <span> </span> <span class="date">26 Jun 2020 at 04:17</span></td>
      </tr>
      <tr>
       <td></td>
       <td><span>I have different elements which are doing the exact same thing but of different classes.</span> <span> - </span> <span class="display-name">Kaneki</span> <span> </span> <span class="date">26 Jun 2020 at 04:19</span></td>
      </tr>
     </tbody>
    </table>
   </div>
   <br>
   <div>
    <h2 id="answer_1"><span>Answer 1</span> <span class="arrow"> <a href="#answer_2">↓</a> </span></h2>
    <div class="container-metadata">
     <div>
      <span>Score: </span> <span>2</span>
     </div>
     <div>
      <span>Answerer: </span> <span>Andreas</span>
     </div>
     <div>
      <span> Answered: </span> <span>26 Jun 2020 at 04:43</span>
     </div>
    </div>
    <div>
     <p>From <a href="https://stackoverflow.com/questions/62587952/moving-from-one-catch-to-another-try-catch-block#comment110683152_62587988">comment</a>:</p>
     <blockquote>
      <p>I have 4 Elements for which I need to create try catch blocks, i.e. 4 try catch blocks</p>
     </blockquote>
     <p>To try 4 different blocks of code, you'd write it like this:</p>
     <pre class="lang-java prettyprint-override"><code>try {
    // Attempt 1
} catch (Exception e1) {
    try {
        // Attempt 2
    } catch (Exception e2) {
        try {
            // Attempt 3
        } catch (Exception e3) {
            try {
                // Attempt 4
            } catch (Exception e4) {
                RuntimeException e = new RuntimeException("All attempts failed", e1);
                e.addSuppressed(e2);
                e.addSuppressed(e3);
                e.addSuppressed(e4);
                throw e;
            }
        }
    }
}
</code></pre>
    </div>
    <hr>
    <div class="comment">
     <table>
     </table>
    </div>
   </div>
   <div>
    <h2 id="answer_2"><span>Answer 2</span> <span class="arrow"> <a href="#answer_1">↑</a> </span> <span class="arrow"> <a href="#answer_3">↓</a> </span></h2>
    <div class="container-metadata">
     <div>
      <span>Score: </span> <span>1</span>
     </div>
     <div>
      <span>Answerer: </span> <span>Avinash tiwari</span>
     </div>
     <div>
      <span> Answered: </span> <span>26 Jun 2020 at 04:21</span>
     </div>
    </div>
    <div>
     <p>You can add finally block below the try catch , finally will run even if some error found in try block . You can look on below code snippet and can replace your code where method1 will have all the code that you want to run in first block and so on</p>
     <pre><code> try {
        method1();
    }finally {
        try {
            method2();
        }finally {
            try {
                method3();
            }finally {
                try {
                    method4();
                }finally {

                }
            }   
        }
    }
</code></pre>
    </div>
    <hr>
    <div class="comment">
     <table>
      <tbody>
       <tr>
        <td></td>
        <td><span>I have 4 Elements for which I need to create try catch blocks, i.e. 4 try catch blocks</span> <span> - </span> <span class="display-name">Kaneki</span> <span> </span> <span class="date">26 Jun 2020 at 04:29</span></td>
       </tr>
       <tr>
        <td></td>
        <td><span>If you have 4 blocks and if that will be constant then you need to manage different try catch for each block</span> <span> - </span> <span class="display-name">Avinash tiwari</span> <span> </span> <span class="date">26 Jun 2020 at 04:33</span></td>
       </tr>
       <tr>
        <td></td>
        <td><span>Everything is happening inside the doInBackground method. All the 4 elements are part of it. So same block.</span> <span> - </span> <span class="display-name">Kaneki</span> <span> </span> <span class="date">26 Jun 2020 at 04:40</span></td>
       </tr>
       <tr>
        <td></td>
        <td><span>Please check the updated answer , you need to move block1 code in method1, block2 code in method2 and so on. You can use catch block also above finally block to print exception</span> <span> - </span> <span class="display-name">Avinash tiwari</span> <span> </span> <span class="date">26 Jun 2020 at 04:52</span></td>
       </tr>
       <tr>
        <td></td>
        <td><span>Let us <a href="https://chat.stackoverflow.com/rooms/216692/discussion-between-avinash-tiwari-and-kaneki">continue this discussion in chat</a>.</span> <span> - </span> <span class="display-name">Avinash tiwari</span> <span> </span> <span class="date">26 Jun 2020 at 04:57</span></td>
       </tr>
      </tbody>
     </table>
    </div>
   </div>
   <div>
    <h2 id="answer_3"><span>Answer 3</span> <span class="arrow"> <a href="#answer_2">↑</a> </span> <span class="arrow"> <a href="#answer_4">↓</a> </span></h2>
    <div class="container-metadata">
     <div>
      <span>Score: </span> <span>0</span>
     </div>
     <div>
      <span>Answerer: </span> <span>Sidhant</span>
     </div>
     <div>
      <span> Answered: </span> <span>26 Jun 2020 at 04:43</span>
     </div>
    </div>
    <div>
     <pre><code>public void tryBlock() {
    try {
        // Enter Your Code;
    } catch (Exception e) {
        System.out.println(" exception :: " + e);
        tryBlock();
    }
}
</code></pre>
    </div>
    <hr>
    <div class="comment">
     <table>
      <tbody>
       <tr>
        <td>1</td>
        <td><span>Probably something like this is the answer, but some comments in addition to the code would be nice. Probably tryBlock needs some parameters as the second call is using "different classes".</span> <span> - </span> <span class="display-name">Tapani</span> <span> </span> <span class="date">26 Jun 2020 at 12:35</span></td>
       </tr>
      </tbody>
     </table>
    </div>
   </div>
   <div>
    <h2 id="answer_4"><span>Answer 4</span> <span class="arrow"> <a href="#answer_3">↑</a> </span> <span class="arrow"> <a href="#answer_5">↓</a> </span></h2>
    <div class="container-metadata">
     <div>
      <span>Score: </span> <span>0</span>
     </div>
     <div>
      <span>Answerer: </span> <span>Subash Hamal</span>
     </div>
     <div>
      <span> Answered: </span> <span>26 Jun 2020 at 05:27</span>
     </div>
    </div>
    <div>
     <p>You can use simple try with multiple catch. Like :</p>
     <pre><code>try{
    something;
}catch(exception){
    Exception2;
}
catch(exception2){
    Exception2;
}


Or you can use nested try-catch like:

try{
  statement
  try{
      statement
      try{
          something
      }catch(exception1){
          Exception
      }
    }catch(exception2){
         Exception2
    }
  }catch(Exception3){
       Exception3
  }  
</code></pre>
    </div>
    <hr>
    <div class="comment">
     <table>
     </table>
    </div>
   </div>
   <div>
    <h2 id="answer_5"><span>Answer 5</span> <span class="arrow"> <a href="#answer_4">↑</a> </span></h2>
    <div class="container-metadata">
     <div>
      <span>Score: </span> <span>0</span>
     </div>
     <div>
      <span>Answerer: </span> <span>Jimmy</span>
     </div>
     <div>
      <span> Answered: </span> <span>26 Jun 2020 at 09:39</span>
     </div>
    </div>
    <div>
     <p>@Kaneki If you want to use nested try-catch block execution in such a way that if operation is not perform inside try block then operation would be in catch block</p>
     <p>For this you have to test for the condition whether you want to forward in catch block or not.</p>
     <p><strong>Note : To forward from <em>Try block</em> to <em>Catch block</em> you have to throw Exception from <em>Try block</em> which must be catched by <em>Catch block</em>.</strong></p>
     <p>This code may help you as per your requirement:</p>
     <pre><code>public void nestedTryCatch(){
        boolean wantTogoForNext = false;
        try{
            wantTogoForNext = false;
            doSomthing("Attempt-I");
            wantTogoForNext = true;
            if(wantTogoForNext)
                throw new Exception("1");
        }catch(Exception e1){
            try{
                wantTogoForNext = false;
                doSomthing("Attempt-2");
                wantTogoForNext = true;
                if(wantTogoForNext)
                    throw new Exception("2");
            }catch(Exception e2){
                try{
                    wantTogoForNext = false;
                    doSomthing("Attempt-3");
                    wantTogoForNext = true;
                    if(wantTogoForNext)
                        throw new Exception("3");
                }catch(Exception e3){
                    try{
                        wantTogoForNext = false;
                        doSomthing("Attempt-4");
                        wantTogoForNext = true;
                        if(wantTogoForNext)
                            throw new Exception("4");
                    }catch(Exception e4){
                        
                        try{
                            wantTogoForNext = false;
                            doSomthing("Attempt-5");
                            wantTogoForNext = true;
                            if(wantTogoForNext)
                                throw new Exception("5");
                        }catch(Exception e5){               
                            
                            doSomthing("Attempt-Last");
                            
                        }   
                    }
                }
            }
        }
    }



public void doSomthing(String attempt){
        System.out.println(attempt);
    }
</code></pre>
    </div>
    <hr>
    <div class="comment">
     <table>
     </table>
    </div>
   </div>
   <br>
   <div>
    <span>License: </span> <a href="https://creativecommons.org/licenses/by-sa/4.0/">CC BY-SA 4.0</a> <span>by </span> <a href="https://stackoverflow.com/legal/terms-of-service#licensing">Stack Overflow Inc.</a>
   </div>
  </div>
 </body>
</html>